WT-6302 move format operations tracing into WiredTiger log files

keithbostic opened the pull request.

keithbostic commented:
@sueloverso, so far nobody has objected to this change, would you be willing to do some initial review?

It's not quite complete, the last chunk of work is to connect verbose logging to the WiredTiger logs, but there's some interesting changes in logging I would like to get approved before we do the detailed review, just to make sure I've not messed up the world.

agorrod commented on test/format/format.h:
```diff
@@ -95,16 +97,18 @@ typedef struct {
+    bool log_all;              /* log all operations  */
+    bool log_local;            /* log to the primary database */
+    WT_CONNECTION *oplog_conn; /* optional logging database */
+    WT_SESSION *oplog_session;
```
can we call this something other than oplog? MongoDB has an oplog and many internal WiredTiger testing applications simulate the oplog behavior and call that "oplog". Using a different name here will help preserve my sanity.

keithbostic replied:
@agorrod, 
> can we call this something other than oplog?

Good point, I'll switch to "trace".

keithbostic commented:
@sueloverso, @agorrod, this one is ready for review.

Sue, Alex decided to leave the work to connect verbose logging to the WiredTiger logs for another ticket, so all references to that change should have been removed from this branch.

I've removed the `-r` option entirely. That option tried to replay runs using the same random number generation as a previous run. In short, the option hasn't worked well in awhile, I don't think anyone is using it, and it slows things down. I'll check to confirm noone is using it before I merge.

I also changed the `-l` option to `-t`, to match the name "trace".

sueloverso commented on src/utilities/util_printlog.c:
```diff
@@ -28,11 +28,14 @@ util_printlog(WT_SESSION *session, int argc, char *argv[])
             break;
+        case 'm': /* messages only */
+            LF_SET(WT_TXN_PRINTLOG_MSG);
+            break;
```
Not related exactly but while in the neighborhood I've also thought it would be useful to have a start/end designation to printlog so that one could only print a subset of the log.

keithbostic replied:
In general I agree, something that would allow us to specify by LSN and/or timestamp. I've created a ticket WT-6309 to track the change.

sueloverso commented on test/format/Makefile.am:
```diff
@@ -4,8 +4,8 @@ AM_CPPFLAGS +=-I$(top_srcdir)/test/utility
 t_SOURCES =\
-	backup.c bulk.c checkpoint.c compact.c config.c config_compat.c kv.c ops.c random.c \
-	rebalance.c salvage.c snap.c t.c util.c wts.c
+	backup.c bulk.c checkpoint.c compact.c config.c config_compat.c kv.c trace.c ops.c \
```
Alphabetize `trace.c`

keithbostic replied:
Fixed in dbc3c6b92

sueloverso commented on test/format/format.i:
```diff
@@ -236,3 +231,38 @@ lock_writeunlock(WT_SESSION *session, RWLOCK *lock)
+            trace_worker(tinfo->trace, "[%" PRIuMAX ":%" PRIuMAX "][%s] " fmt,              \
+              (uintmax_t)__ts.tv_sec, (uintmax_t)__ts.tv_nsec / WT_THOUSAND, tinfo->tidbuf, \
+              __VA_ARGS__);                                                                 \
+        }                                                                                   \
```
AFAICT the only difference between these two is getting the session. How about refactoring into a macro and have each of these just pass in the session. Then there's only one format string, args and caller to `trace_worker`.

keithbostic replied:
They also use different thread-ID buffers.

I went ahead and cleaned up some stuff in 7ebeaa50b, and now I'm no longer using `WT_SESSION.log_vprintf()`. That said, I really think we should offer both, it's standard everywhere as far as I know.

sueloverso commented on test/format/t.c:
```diff
@@ -237,18 +240,6 @@ main(int argc, char *argv[])
 
-    /*
-     * Multithreaded runs can be replayed: it's useful and we'll get the configuration correct.
-     * Obviously the order of operations changes, warn the user.
```
I'm not sure how the removal of the replay functionality fits into adding format logging to WT log files. Is this a gratuitous change?

keithbostic replied:
Yes and no. I've been thinking about removing `replay` for awhile, but when I took a run at making it work in the new format structure, it was hard to get right. Given I don't think it's worth keeping, I didn't see any reason to do that work.

sueloverso commented on test/format/t.c:
```diff
@@ -375,19 +365,19 @@ static void
       "\t-c read test program configuration from a file (default 'CONFIG')\n"
       "\t-h home directory (default 'RUNDIR')\n"
-      "\t-l log operations to a file\n"
+      "\t-L all|local\n"
```
This `-T` and `-t` in the code, I think.

keithbostic replied:
Fixed in a698f4104.

sueloverso commented on test/format/trace.c:
```diff
@@ -0,0 +1,132 @@
+        len = strlen(g.home) * strlen(TRACE_DIR) + 10;
+        p = dmalloc(len);
+        testutil_check(__wt_snprintf(p, len, "%s/%s", g.home, TRACE_DIR));
+        config = "create,log=(enabled,archive),debug_mode=(log_retain=10)";
```
`log_retain=10` is not consistent with the comment about keeping the last 20. Probably remove the number from the comment and make it general.

sueloverso replied:
Also, just to point it out, the default file size is 100Mb, which you're not overriding. Given this is a separate database with no other log operations, you're retaining 1Gb of messages. That is still a lot to wade through.

keithbostic replied:
Fixed in 3ac7364.

sueloverso commented on test/format/trace.c:
```diff
@@ -0,0 +1,132 @@
+    testutil_check(conn->open_session(conn, NULL, NULL, &session));
+
+    g.trace_conn = conn;
+    g.trace_session = session;
```
Ah, I understand now. That is why you don't want LSNs or they wouldn't be useful.

keithbostic replied:
Not sure I'm following, but if you're good, I'm good.

sueloverso replied:
I was thinking that LSNs would be useful so that you could look at specific operations around that LSN but since this is a separate connection, there is no way to connect something in the "real" connection/log with anything in this trace log. At the time I wrote that comment, I thought it was all intermingled in the log. That still might be an okay idea. I don't think that writing messages slows down logging or the test much.

keithbostic replied:
That is my intent: there's a flag in this change, `-L local` that puts format traces into the primary database log instead of a separate database. Once we do WT-6304 (put verbose messages into the log), then there's a single set of logs that contains verbose messages, format messages, and normal log records, so looking at the activity around a particular LSN (or timestamp), makes a lot of sense.

We could also hook it up so verbose messages go into the trace database I'm creating for format, but I think that's generally less interesting, I agree with you that adding messages to the log won't matter much.

sueloverso commented on src/txn/txn_log.c:
```diff
@@ -660,6 +660,13 @@ __txn_printlog(WT_SESSION_IMPL *session, WT_ITEM *rawrec, WT_LSN *lsnp, WT_LSN *
     /* First, peek at the log record type. */
     WT_RET(__wt_logrec_read(session, &p, end, &rectype));
 
+    if (F_ISSET(args, WT_TXN_PRINTLOG_MSG)) {
```
I'd add a comment here about how this flag is used. It is not obvious to reason about without context when looking at printing out log message records further down.

Also, I think it would be useful to know the LSN for the messages so that one can easily find the log file and see the records and activity around that time.

keithbostic replied:
I added a comment in 59290b1b1.

I don't want to add the LSN (there are timestamps), but if we ever offer a way to print out a range of log records, I think it's easier to make that argument.

sueloverso commented on src/conn/conn_log.c:
```diff
@@ -386,18 +386,26 @@ __log_archive_once(WT_SESSION_IMPL *session, uint32_t backup_file)
-         * Figure out the minimum log file to archive. Use the LSN in the debugging array if
-         * necessary.
+         * If we're performing checkpoints apply the retain value as a maximum, increasing the
+         * number the number of we keep. If not performing checkpoints, let it be an absolute number
```
`the number the number of we keep`.

keithbostic replied:
Fixed in 91c511668.

keithbostic commented on src/include/wiredtiger.in:
```diff
@@ -1549,6 +1549,18 @@ struct __wt_session {
+	 * @param ap a reference to a variable list of arguments
+	 * @errors
+	 */
+	int __F(log_vprintf)(WT_SESSION *session, const char *format, va_list ap);
```
This seems reasonable to me, there really aren't any variadic argument APIs that don't include a `va_list *` variant.

sueloverso replied:
Now that you avoided using the API, it has no users. I mean, `log_printf` itself is not widely used already and has been sufficient for so many years. Adding this here clutters up the change, but a new API also means you need changes to documentation (`cursor-log.dox`), the python interface, java interface, example programs, etc to be real and complete with it. While I can agree with your sentiment about the variadic argument, maybe a tech debt ticket something solely focused on the API is the best route?

keithbostic commented on dist/api_data.py:
```diff
@@ -522,6 +522,9 @@ def __ge__(self, other):
             type='boolean'),
+        Config('log_retain', '0', r'''
+            adjust log archiving to retain at least this number of log files''',
+            min='0', max='1024'),
```
Switched to `log_retention`, pushed in 0d74f44.

keithbostic commented on src/conn/conn_log.c:
```diff
@@ -386,18 +386,26 @@ __log_archive_once(WT_SESSION_IMPL *session, uint32_t backup_file)
+    /* Adjust the number of log files to retain based on debugging options. */
+    if (conn->debug_ckpt_cnt != 0)
+        min_lognum = WT_MIN(conn->debug_ckpt[conn->debug_ckpt_cnt - 1].l.file, min_lognum);
+    if (conn->debug_log_retain != 0) {
```
Switched to `debug_log_cnt`, pushed in 0d74f44.

keithbostic commented on src/include/txn.i:
```diff
@@ -948,9 +948,9 @@ __wt_txn_read(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt, WT_ITEM *key, uint
-     * and should return "not found", except for history store scan during rollback to stable and
-     * when we are told to ignore non-globally visible tombstones.
+     * not prepared and it is visible to our txn it means we've just spotted a tombstone and should
+     * return "not found", except for history store scan during rollback to stable and when we are
```
Fixed in 4796deb27.

sueloverso commented on src/session/session_api.c:
```diff
@@ -813,7 +813,7 @@ __session_log_printf(WT_SESSION *wt_session, const char *fmt, ...)
 
     session = (WT_SESSION_IMPL *)wt_session;
-    SESSION_API_CALL_PREPARE_NOT_ALLOWED_NOCONF(session, log_printf);
+    SESSION_API_CALL_PREPARE_ALLOWED_NOCONF(session, log_printf);
```
This change is breaking some of the unit tests.

keithbostic replied:
Thanks, fixed in 7cf79d0.

keithbostic commented on src/conn/conn_log.c:
```diff
@@ -386,18 +386,26 @@ __log_archive_once(WT_SESSION_IMPL *session, uint32_t backup_file)
+        if ((conn->debug_log_cnt + 1) >= log->fileid)
+            return (0);
+        if (log->ckpt_lsn.l.file == 1 && log->ckpt_lsn.l.offset == 0)
+            min_lognum = log->fileid - (conn->debug_log_cnt + 1);
```
I added comments to the code and the documentation in 6b4480ca6.

keithbostic commented:
@sueloverso, I removed `WT_SESSION.log_vprintf()` in 7cf79d020.

agorrod commented on dist/api_data.py:
```diff
@@ -522,6 +522,13 @@ def __ge__(self, other):
+            done and the number of log files exceeds the configured value. As WiredTiger cannot
+            detect the difference between a system that has not yet checkpointed and one that
+            will never checkpoint, it might discard log files before any checkpoint is done.)''',
+            min='0', max='1024'),
```
Does '0' mean keep all log files, or keep none?

keithbostic replied:
I think it should mean the value is simply ignored, so the default rules apply, whatever they may be. Documented in 2c05d4728.
